source/UF/Pushouts.lagda

```diff
@@ -0,0 +1,418 @@
+Pushout-Induction-Principle {_} {_} {_} {_} {_} {A} {B} {C} S f g (i , j , G) P 
+ = (l : (a : A) → P (i a))
+ → (r : (b : B) → P (j b))
+ → ((c : C) → transport P (G c) (l (f c)) ＝ r (g c))
+ → (x : S) → P x
+
+Pushout-Computation-Rule₁
+ : {A : 𝓤  ̇} {B : 𝓥  ̇} {C : 𝓦  ̇} (S : 𝓤'  ̇) 
+   (f : C → A) (g : C → B) (s : cocone f g S) (P : S → 𝓣  ̇)
+ → Pushout-Induction-Principle S f g s P
+ → 𝓤 ⊔ 𝓥 ⊔ 𝓦 ⊔ 𝓣  ̇
+Pushout-Computation-Rule₁
+ {_} {_} {_} {_} {_} {A} {B} {C} S f g (i , j , G) P S-ind
+ = (l : (a : A) → P (i a))
+ → (r : (b : B) → P (j b))
+ → (H : (c : C) → transport P (G c) (l (f c)) ＝ r (g c))
+ → (a : A)
+ → S-ind l r H (i a) ＝ l a
+
+Pushout-Computation-Rule₂
+ : {A : 𝓤  ̇} {B : 𝓥  ̇} {C : 𝓦  ̇} (S : 𝓤'  ̇)
+   (f : C → A) (g : C → B) (s : cocone f g S) (P : S → 𝓣  ̇)
+ → Pushout-Induction-Principle S f g s P
+ → 𝓤 ⊔ 𝓥 ⊔ 𝓦 ⊔ 𝓣  ̇
+Pushout-Computation-Rule₂
+ {_} {_} {_} {_} {_} {A} {B} {C} S f g (i , j , G) P S-ind
+ = (l : (a : A) → P (i a))
+ → (r : (b : B) → P (j b))
+ → (H : (c : C) → transport P (G c) (l (f c)) ＝ r (g c))
+ → (b : B)
+ → S-ind l r H (j b) ＝ r b
+
+Pushout-Computation-Rule₃
+ : {A : 𝓤  ̇} {B : 𝓥  ̇} {C : 𝓦  ̇} (S : 𝓤'  ̇) 
+   (f : C → A) (g : C → B) (s : cocone f g S) (P : S → 𝓣  ̇)
+   (S-ind : Pushout-Induction-Principle S f g s P)
+ → Pushout-Computation-Rule₁ S f g s P S-ind
+ → Pushout-Computation-Rule₂ S f g s P S-ind
+ → 𝓤 ⊔ 𝓥 ⊔ 𝓦 ⊔ 𝓣  ̇
+Pushout-Computation-Rule₃
+ {_} {_} {_} {_} {_}{A} {B} {C} S f g (i , j , G) P S-ind S-comp₁ S-comp₂
+ = (l : (a : A) → P (i a))
+ → (r : (b : B) → P (j b))
+ → (H : (c : C) → transport P (G c) (l (f c)) ＝ r (g c))
+ → (c : C)
+ → apd (S-ind l r H) (G c) ∙ S-comp₂ l r H (g c)
+ ＝ ap (transport P (G c)) (S-comp₁ l r H (f c)) ∙ H c
+
+\end{code}
+
+The following are logically equivalent (which is an instance of Sojakova's result):
+
+1) The dependent universal property
+2) The induction principle with propositional computation rules
+3) The recursion principle with propositional computation rules and the
+   uniqueness principle
+4) The universal property.
+
+Below we will derive 2), 3) and 4) from the seemingly strongest assumption 1).
+In a later file we will attempt to derive 1), 2) and 3) from the seemingly weakest
+assumption 4) (this is a work in progress.)
+
+\begin{code}
+
+record pushouts-exist {A : 𝓤  ̇} {B : 𝓥  ̇} {C : 𝓦  ̇} (f : C → A) (g : C → B) : 𝓤ω
+ where
+ field
+  pushout : 𝓤 ⊔ 𝓥 ⊔ 𝓦  ̇
+  inll : A → pushout 
+  inrr : B → pushout 
+  glue : (c : C) → inll (f c) ＝ inrr (g c)
+  pushout-dependent-universal-property
+   : {P : pushout → 𝓣  ̇}
+   → Pushout-Dependent-Universal-Property pushout f g (inll , inrr , glue) P
+
+\end{code}
+
+We need to unpack all the information from the dependent universal property (i.e.
+we extract the fact that the fiber of the canonical map is contractible).
+
+\begin{code}
+
+ pushout-cocone : cocone f g pushout
+ pushout-cocone = (inll , inrr , glue)
+
+ pushout-fiber-is-singleton
+  : {P : pushout →  𝓣'  ̇}
```

Comment:
Module parameters
```
{P : pushout →  𝓣'}
(t : dependent-cocone f g pushout pushout-cocone P)
```
would clean this up. Also note that I removed the spacing after `𝓣'` and before `}`.